packages/workspace/src/browser/workspace-service.ts

```diff
@@ -55,30 +67,174 @@ export class WorkspaceService implements FrontendApplicationContribution {
     @inject(WorkspacePreferences)
     protected preferences: WorkspacePreferences;
 
+    @inject(ResourceProvider)
+    protected readonly resourceProvider: ResourceProvider;
+
     @postConstruct()
     protected async init(): Promise<void> {
-        const rootUri = await this.server.getRoot();
-        this._root = await this.toValidRoot(rootUri);
+        this._workspaceRootChange = {
+            root: undefined,
+            roots: [],
+            workspaceConfigFile: undefined
+        };
+        await this.setupRoots();
+        this.updateTitle();
+        this.deferredWorkspaceRootChange.resolve(this._workspaceRootChange);
+        if (this._workspaceRootChange.workspaceConfigFile) {
+            await this.watchRootConfigFile();
+        } else {
+            const configUri = await this.getWorkspaceConfigFileUri();
+            if (configUri) {
+                this.watcher.onFilesChanged(async event => {
```

Comment:
This looks expensive, but we need to handle the scenario where the `roots.json` does not exist in the first place. Or maybe we could use a timer as suggested in https://github.com/Axosoft/nsfw/issues/45. Not sure which is better.